docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,451 @@
+
+### User Stories
+
+#### Story 1
+
+As a developer or Cluster operator, I would like Cluster add-on providers to automatically include or install the package management tool.
+
+#### Story 2
+
+As a developer or Cluster operator, I would like Cluster add-on providers to expose configurations native to the package management tool to install Cluster add-ons into the newly created Workload Cluster.
+
+#### Story 3
+
+As a developer or Cluster operator, I would like Cluster add-on providers to leverage existing capabilities of the package management tool, such as identifying the proper version of Cluster add-ons package to install onto the workload Cluster.
+
+#### Story 4
+
+As a developer or Cluster operator, I would like Cluster add-on providers to use information available in the Management Cluster (e.g. service or pod CIDR on the Cluster object) for customizing add-ons configurations specifically for the newly created workload Cluster.
+
+#### Story 5
+
+As a developer or Cluster operator, I would like Cluster add-on providers to update add-ons configurations when information available in the Management Cluster (e.g add-on configuration) changes.
+
+### ClusterAddonProvider Functionality
+
+A ClusterAddonProvider is a new component for bringing add-ons into the Cluster API lifecycle. It serves as a general solution to Cluster API Add-ons with multiple concrete implementations, similar to the relationship between Cluster API and infrastructure providers. It aims to do so handle add-on lifecycle management in a holistic way by providing the following capabilities:
+
+- Provide or install a package management tool
+- Determine which add-ons to install or uninstall on each Cluster
+- Provide a default configuration for each add-on
+- Configure an add-on by resolving or inherit information from the Cluster configuration (i.e. service or pod CIDR for CNI)
+- Install add-ons during cluster creation (immediately after the API server is installed). This operation usually brings in some additional requirements such as: 
+  - Resolve the add-on version for the Cluster’s Kubernetes version
+  - Wait for the add-on to be actually running
+- Provide a mechanism for upgrading an add-on before, during, or after the Cluster upgrade and cleaning up the previous version
+- Support for out-of band upgrades for add-ons not linked to the Cluster lifecycle, i.e. CVE fix
+- Configure pre and post installation steps for each add-on, thus allowing to manage migrations/operations such as:
+  - CPI migration from in-tree to out-of-tree
+  - CoreDNS configuration migration from one version to another
+- Orchestrate add-on deletion before Cluster deletion
```

Comment:
I was discussing with @CecileRobertMichon about this point. Is there a use case for why we'd want to delete the add-on if the Cluster is getting delete anyway? Or in other words would we ever expect users to `helm delete` their charts before deleting a Cluster?

Reply:
I think as long as they are stateless or their data is pushed/pulled to somewhere else anyway we shouldn't care. I think this applies to 99% of the cases we currently consider (including stuff like log shipping and metrics collection).

Reply:
Gotcha, I guess my question then is do we want to explicitly set an owner ref from HelmReleaseProxy or instead reword this point a bit? If we do set an ownerRef, then that forces it to be in the same namespace as the Cluster.

Reply:
We should probably decide first if the CRDs are namespaced or cluster-wide. This seems to be a major point and it has influence on a lot of the currently open discussions.

Reply:
I think adding a namespace match in addition to the cluster label match perhaps makes things overly complicated. This is also a feature we can add later if folks decide they want to compose cluster label key/value + namespace.

So my vote would be to omit the cluster namespace and default to "cluster-wide".